### DISCUSSION TITLE
Optimization in DHNx

### DISCUSSION THREAD

**Discussion by gitfeldhase** [oemof/DHNx#100] [Q&A] [answered]
Hi,

I am wondering how the optimization workflow of DHNx works.
I saw a discussion in the following Issue and added some questions there.
[https://github.com/oemof/DHNx/issues/66](https://github.com/oemof/DHNx/issues/66)

I am aware that in the future, the discussions board should be used but would be glad if somebody could help with this issue.

Thanks a lot for any hints.

Best regards
Lars

**Comment by joroeder** [comment, marked as answer]
Maybe, lets simply continue the discussion of #66 here, as I think the investment input table is something that needs to made clearer.

So, thank you for your comments so far.

Here, you will find the pipes.csv:
[pipes.csv](https://github.com/oemof/DHNx/files/10176516/pipes.csv)

I also added an example that shows the pre-calculation of these parameters based on given district heating pipeline data for each DN number:
[hydraulic_precalculation.py](https://github.com/oemof/DHNx/blob/dev/examples/optimisation/precalculation/hydraulic_precalculation.py)

In the documentation, I have also tried to describe the pipelines input table here: [https://dhnx.readthedocs.io/en/latest/optimization_models.html#network-pipes-csv](https://dhnx.readthedocs.io/en/latest/optimization_models.html#network-pipes-csv)

But maybe, altogether, the input needs to be revised for better usage. So, feedback is welcome, this can be improved in future.

Some general remarks:

* forward and return pipes are summarized in a single DHS pipe trench with a transport capacity
* the transport capacity of a certain DN number results from a pre-calculation, and the transport capacity is the quantity to which to costs and losses are related.
* each row of the `network/pipes.csv` creates an `Heatpipeline` object for each of the fork-fork, fork-consumer connection. That means each row can be used as e.g. different linearized segments of the cost-curve of the district heating pipelines
* the `nonconvex` attribute is important if you want to consider fixed losses and costs, that do not depend on the pipe capacity -- and this is necessary to approximate the cost-curve of DHS pipelines. Please see also: [https://oemof-solph.readthedocs.io/en/v0.4.4/usage.html#using-the-investment-mode](https://oemof-solph.readthedocs.io/en/v0.4.4/usage.html#using-the-investment-mode)

I hope that helps!

**Reply by gitfeldhase** [reply]
> > It seems that for the routing problem the oemof model uses only the shortest path for every consumer to the heat source instead of the cheapest investment. ![Screenshot from 2022-11-28 16-56-51](https://user-images.githubusercontent.com/106976983/204323333-bc829a79-eec8-445b-b5f5-aea088a05a17.png)
> 
> It is based on a cost-minization, so in your case, the reason probably is that you do not consider fixed cost, as nonconvex is false? So the connection of the building as in your results is the cheapest way. Try to set `nonconvex` to 1 and add `fix_costs`.
> 
> > I get this warning in my log: `There is NO investment in both directions at thefollowing pipes for " pipe-typ-A "`
> > Can you help me out what the log means and what am I missing about the routing problem?
> 
> That is a good question, could you send the example?

Thanks a lot for the answer!

Here is the data of the example:

[invest_data.zip](https://github.com/oemof/DHNx/files/10184209/invest_data.zip)

Apparently I get the same result with nonconvex = 1 and additional fix costs, no matter how high I choose the additional fix costs.

##### EDIT #####
I used the script of this example to run the optimization and got the correct network typology!
https://github.com/oemof/DHNx/blob/dev/examples/optimisation/minimal_network/minimal_network.py
It works great, the mistake must have been somewhere in my script when handing in the invest data I guess after preprocessing my geodata.

Thanks for the support, it really helped a lot!
![image](https://user-images.githubusercontent.com/106976983/206424459-ad6c4715-a278-48a9-90f7-68f28b3be49a.png)

Lars

**Reply by gitfeldhase** [reply]
Hi, sorry i checked it again and just copied my example into the optimization script in the example and it works now!!!
Something must have been wrong with my setup.

Thanks anyways! I really learned a lot with this conversation and got some really good knowledge about oemof.solph.

**Reply by joroeder** [reply]
Great! :)
